utils/python/CIME/SystemTests/system_tests_common.py

```diff
@@ -48,70 +51,142 @@ def __init__(self, case, expected=None):
         """
-        self._case = case
-        self._caseroot = case.get_value("CASEROOT")
-
-    def _run(self, suffix="base", coupler_log_path=None, st_archive=False):
+        success = True
+        start_time = time.time()
         try:
-            success = case_run(self._case)
-            if success and st_archive:
-                success = case_st_archive(self._case)
+            expect(self._test_status.get_status(MODEL_BUILD_PHASE) == TEST_PASS_STATUS,
+                   "Model was not built!")
+            with self._test_status:
+                self._test_status.set_status(RUN_PHASE, TEST_PENDING_STATUS)
 
-            if success and self._coupler_log_indicates_run_complete(coupler_log_path):
-                self.pass_test()
-            else:
-                success = False
-                self.fail_test()
+            success = self.run_phase()
+            if not isinstance(success, bool):
+                logger.warning("build_phase did not return a bool, assuming PASS!")
+                success = True
+
+            if success:
+                if self._case.get_value("GENERATE_BASELINE"):
+                    self._generate_baseline()
+
+                if self._case.get_value("COMPARE_BASELINE"):
+                    self._compare_baseline()
+
+                self._check_for_memleak()
 
-            if success and suffix is not None:
-                self._component_compare_move(suffix)
         except:
-            # An exception must not prevent the TestStatus file from
-            # being marked FAIL
             success = False
-            self.fail_test()
             logger.warning("Exception during run: %s" % (sys.exc_info()[1]))
 
+        # Always try to report
+        self.report()
+
+        # Writing the run status should be the very last thing due to wait_for_tests
+        time_taken = time.time() - start_time
+        status = TEST_PASS_STATUS if success else TEST_FAIL_STATUS
+        with self._test_status:
+            self._test_status.set_status(RUN_PHASE, status, comments=("TIME=%d" % int(time_taken)))
+
         return success
 
-    def __del__(self):
-        if self._runstatus is not None:
-            test_status = os.path.join(self._orig_caseroot, "TestStatus")
-            with open(test_status, 'r') as f:
-                teststatusfile = f.read()
-            li = teststatusfile.rsplit('PEND', 1)
-            teststatusfile = self._runstatus.join(li)
-            total_time = int(time.time() - self._start_time)
-            with open(test_status, 'w') as f:
-                f.write(teststatusfile)
-                f.write("COMMENT TIME %d\n" % total_time)
+    def run_phase(self):
+        """
+        This is the default run phase implementation, it just does an individual run.
+        This is the subclass' extension point if they need to define a custom run phase.
+
+        PLEASE THROW AN EXCEPTION OR RETURN FALSE ON FAIL
+        """
+        return self.run_indv()
```

Comment:
@billsacks : I wanted to give the test developer a choice of returning False or throwing an exception. Currently, I changed every test to just rely on exceptions since it's a little simpler. The vision of run_indv is that users would use try/except if they wanted to handle the error instead of letting it escape. That said, if you think that this is confusing, it would be easy to change.